go/vt/vttablet/tabletserver/tx_prep_pool.go

```diff
@@ -169,3 +172,25 @@ func (pp *TxPreparedPool) FetchAllForRollback() []*StatefulConnection {
 	pp.reserved = make(map[string]error)
 	return conns
 }
+
+func (pp *TxPreparedPool) IsEmpty(tableName string) bool {
+	pp.mu.Lock()
+	defer pp.mu.Unlock()
+	if !pp.twoPCEnabled {
+		return true
+	}
+	// If the pool is shutdown, we do not know the correct state of prepared transactions.
+	if !pp.open {
+		return false
+	}
+	for _, connection := range pp.conns {
+		for _, query := range connection.txProps.Queries {
+			for _, table := range query.Tables {
+				if table == tableName {
+					return false
+				}
+			}
+		}
+	}
+	return true
+}
```

Comment:
Let's discuss the race condition again: this function is called when query buffering is already in place, which technically prevents a new prepared transaction. And given we also intentionnal `sleep(100*time.Milisecond)` in between, I don't see how realistically this might happen, but:

- Say a prepared transaction gets as far as this line:
https://github.com/vitessio/vitess/pull/16585/files#diff-a1efa75583f31e230076c9385a551deb725040bbbf85085833be7c14a7a18055R102
- Say Online DDL cut-over begins and starts buffering.
- The prepared transaction above proceeds a bit further.
- When Online DDL invokes `IsEmpty()`, will it see the prepared transaction above? If yes, fine. If not, then it might kill it.

Reply:
Now we give even prepared pool to get empty in 100ms. This will handle the case a transaction gets added to prepared pool after query rules are added.
Within this 100ms, it will recheck the rules and fail the prepare step, allowing online ddl cutover to proceed.

Reply:
the other check if if if gets past the first query rule check and not added to prepared pool list, then online DDL will kill the connection. This is also handled as we check if the MySQL connection is closed before proceeding.

Reply:
> the other check if if if gets past the first query rule check and not added to prepared pool list

The race condition is what happens if it gets past the first check, then Online DDL doesn't see it in the prepared pool list and proceeds towards the killing function, then the query gets into the prepared pool list, then Online DDL actually finds it and kills it.

Reply:
The list is just a first check point. 
The prepare happens latter own, when we update the database with the redo logs. Till then any failure is accepted